app/cluster/dynamic.go

```diff
@@ -62,33 +74,60 @@ func (d *Dynamic) init() {
 	d.serverStopTimeStat = stats.NewTaggedStat("cluster.server_stop_time", stats.TimerType, tag)
 	d.serverStartCountStat = stats.NewTaggedStat("cluster.server_start_count", stats.CountType, tag)
 	d.serverStopCountStat = stats.NewTaggedStat("cluster.server_stop_count", stats.CountType, tag)
+
+	if d.BackendConfig == nil {
+		d.BackendConfig = backendconfig.DefaultBackendConfig
+	}
 }
 
 func (d *Dynamic) Run(ctx context.Context) error {
-	d.once.Do(func() {
-		d.init()
-	})
-	defer d.Provider.Close()
-	modeChan, err := d.Provider.ServerMode()
-	if err != nil {
-		return err
-	}
+	d.once.Do(d.init)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	serverModeChan := d.Provider.ServerMode(ctx)
+	workspaceIDsChan := d.Provider.WorkspaceIDs(ctx)
+
 	for {
 		select {
 		case <-ctx.Done():
 			if d.currentMode == servermode.NormalMode {
 				d.stop()
 			}
 			return nil
-		case newMode := <-modeChan:
-			d.logger.Infof("Got trigger to change the mode, new mode: %s, old mode: %s", newMode.Mode(), d.currentMode)
-			err = d.handleModeChange(newMode.Mode())
+		case req := <-serverModeChan:
+			if req.Err() != nil {
+				return req.Err()
```

Comment:
Might be a dumb question but where exactly do we handle this error? I see this being propogated till main but never acted upon or logged? Any reason why

Reply:
My understanding is that an error like this should cause this application to stop. By propagating to main we ensure it will stop in a graceful way.